lbs-dynamodb/metadata.c

```diff
@@ -316,6 +322,47 @@ metadata_nextblk_write(struct metadata * M, uint64_t nextblk,
 	return (-1);
 }
 
+/**
+ * metadata_lastblk_read(M):
+ * Return the "lastblk" value.
+ */
+uint64_t
+metadata_lastblk_read(struct metadata * M)
+{
+
+	/* Return currently stored value. */
+	return (M->M_stored.lastblk);
+}
+
+/**
+ * metadata_lastblk_write(M, lastblk, callback, cookie):
+ * Store "lastblk" value.  Invoke ${callback}(${cookie}) on success.
+ */
+int
+metadata_lastblk_write(struct metadata * M, uint64_t lastblk,
+    int (*callback)(void *), void * cookie)
+{
+
+	/* We shouldn't have a callback already. */
+	assert(M->M_latest.callback == NULL);
+
+	/* Record the new value and callback parameters. */
+	M->M_latest.lastblk = lastblk;
+	M->M_latest.callback = callback;
```

Comment:
At this point in the branch, there's no `callback` inside `struct mtuple`.  I see two possible fixes:

1. add a `callback_lastblk` and `cookie_lastblk` to `struct medata`, make this function an almost-copy of `metadata_nextblk_write`, and add other infrastructure as needed.  (Then revert those changes somewhere later in the branch.)
2. or, add a commit before this one which adds the `callback` to `struct mtuple` (as is done later in this branch), possibly also removes `struct metadata -> callback_nextblk` (again, as is done later), then leaves this commit unchanged.

I've gone back and forth a few times, and have half-implemented both solutions with low confidence (particularly since `make test` for this code takes 45 minutes).  At this point, I'd like to get your thoughts about the nicest solution, then focus on that.

Reply:
Oh, I guess I reordered commits in a way which broke them.  Oops.

`struct mtuple.callback` is the same thing as `callback_nextblk`.  Similarly `cookie` and `cookie_nextblk`.

If you want to make this super tidy: Add a new commit which renames `{callback, cookie}_nextblk` to `{callback, cookie}_state` -- they're used by callbacks to `state.c` and there can only be one pending at once.  Then those can move into `struct mtuple` later.